### PR TITLE
Fix auth proxy support

### PR DESCRIPTION
**Pull request by jacobprice808**
Edits utils.py to create a custom extension in order to implement proxies with auth credentials. I've tested it and it is working for my auth proxies. I don't have any non-auth proxies to verify that the original method is still working, so someone might want to test that. Also, I use prettier in VSCode with reformatted existing code when I saved, so sorry about that.

Closes #803

### CODE REVIEW

File: src/undetected_chromedriver/__init__.py [hunk 1]
```diff
@@ -17,7 +17,7 @@
 from __future__ import annotations
 
 
-__version__ = "3.4.6"
+__version__ = "3.2.1"
```
Comment (ilike2burnthing) [inline src/undetected_chromedriver/__init__.py:20]: Wrong version used?

Comment (jacobprice808) [reply src/undetected_chromedriver/__init__.py]: Yeah, I accidentally copied your patch thinking it was a patch on top of @DX37's pull request. I fixed it in my fork but now FlareSolverr isn't working so I need to figure out what happened.

**Comment by jacobprice808** [comment]
#[82f7ff7](https://github.com/FlareSolverr/FlareSolverr/pull/807/commits/82f7ff7d646b559758c9b5414a47e776282b27b8) isn't working. I upgraded undetected chromedriver and applied the patches as per #803. Need to debug. Also, I need to uninstall Prettier, lol.

**Comment by jacobprice808** [comment]
Okay, I fixed it. It's working well for me now. Someone else can verify that it's working on their end. I had patcher.py in the wrong directory. [#0468d92](https://github.com/FlareSolverr/FlareSolverr/pull/807/commits/0468d92c25a8c59e8a6753d6d0acbe273b376eba) should should satisfy #807 and #803.

**Comment by jacobprice808** [comment]
There was a bug that would mess up concurrent requests because they would both be trying to access and modify a dynamically created extension. I fixed it by creating a fresh temporary extension for every initiated session.

**Comment by ilike2burnthing** [comment]
@jacobprice808 so the way I'm testing this is with Jackett (also happens with Prowlarr), it's timing out with the current v3.2.1 build and with your PR build.

Specifically it's trying and failing to find the `Verify you are human` button over and over:

```text
2023-07-12 00:05:06 DEBUG    ReqId 5396 Navigating to... https://idope.se/browse.html
2023-07-12 00:05:07 INFO     ReqId 5396 Challenge detected. Title found: Just a moment...
2023-07-12 00:05:07 DEBUG    ReqId 5396 Waiting for title (attempt 4): Just a moment...
2023-07-12 00:05:18 DEBUG    ReqId 5396 Timeout waiting for selector
2023-07-12 00:05:18 DEBUG    ReqId 5396 Try to find the Cloudflare verify checkbox
2023-07-12 00:05:18 DEBUG    ReqId 5396 Cloudflare verify checkbox found and clicked
2023-07-12 00:05:18 DEBUG    ReqId 5396 Try to find the Cloudflare 'Verify you are human' button
2023-07-12 00:05:18 DEBUG    ReqId 5396 The Cloudflare 'Verify you are human' button not found on the page
```

Are you able to replicate this? A few public indexers to use for this are TorrentQQ, Torrent[CORE], and Idope.

Is FlareSolverr working outside of Jackett for you?

---

Also, while I can run your PR build of FlareSolverr fine the first time, upon trying to relaunch it I get:

```text
C:\Users\WDAGUtilityAccount\Desktop\FlareSolverr-master>python src/flaresolverr.py
2023-07-11 23:49:31 INFO     FlareSolverr 3.2.1
2023-07-11 23:49:31 INFO     Testing web browser installation...
2023-07-11 23:49:31 INFO     Platform: Windows-10-10.0.19041-SP0
2023-07-11 23:49:31 INFO     Chrome / Chromium path: C:\Program Files\Google\Chrome\Application\chrome.exe
2023-07-11 23:49:31 INFO     Chrome / Chromium major version: 114
2023-07-11 23:49:31 INFO     Launching web browser...
Traceback (most recent call last):
  File "C:\Users\WDAGUtilityAccount\Desktop\FlareSolverr-master\src\utils.py", line 319, in get_user_agent
    driver = get_webdriver()
             ^^^^^^^^^^^^^^^
  File "C:\Users\WDAGUtilityAccount\Desktop\FlareSolverr-master\src\utils.py", line 196, in get_webdriver
    shutil.copy(driver.patcher.executable_path, PATCHED_DRIVER_PATH)
  File "C:\Users\WDAGUtilityAccount\AppData\Local\Programs\Python\Python311\Lib\shutil.py", line 419, in copy
    copyfile(src, dst, follow_symlinks=follow_symlinks)
  File "C:\Users\WDAGUtilityAccount\AppData\Local\Programs\Python\Python311\Lib\shutil.py", line 236, in copyfile
    raise SameFileError("{!r} and {!r} are the same file".format(src, dst))
shutil.SameFileError: 'C:\\Users\\WDAGUtilityAccount\\appdata\\roaming\\undetected_chromedriver\\chromedriver.exe' and 'C:\\Users\\WDAGUtilityAccount\\appdata\\roaming\\undetected_chromedriver\\chromedriver.exe' are the same file

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\WDAGUtilityAccount\Desktop\FlareSolverr-master\src\flaresolverr.py", line 102, in <module>
    flaresolverr_service.test_browser_installation()
  File "C:\Users\WDAGUtilityAccount\Desktop\FlareSolverr-master\src\flaresolverr_service.py", line 72, in test_browser_installation
    user_agent = utils.get_user_agent()
                 ^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\WDAGUtilityAccount\Desktop\FlareSolverr-master\src\utils.py", line 323, in get_user_agent
    raise Exception("Error getting browser User-Agent. " + str(e))
Exception: Error getting browser User-Agent. 'C:\\Users\\WDAGUtilityAccount\\appdata\\roaming\\undetected_chromedriver\\chromedriver.exe' and 'C:\\Users\\WDAGUtilityAccount\\appdata\\roaming\\undetected_chromedriver\\chromedriver.exe' are the same file
```

This does not occur with v3.2.1.

---

Lastly, this may be a Jackett or maybe FlareSolverrSharp issue, but it doesn't seem that Jackett's proxy credentials are being passed to FlareSolverr:

```text
2023-07-12 00:11:48 INFO     ReqId 5116 Incoming request => POST /v1 body: {'maxTimeout': 55000, 'cmd': 'request.get', 'url': 'https://idope.se/browse.html', 'proxy': {'url': 'socks5://stockholm.se.socks.nordhold.net:1080'}}
2023-07-12 00:11:48 DEBUG    ReqId 5116 Launching web browser...
2023-07-12 00:11:48 ERROR    ReqId 5116 Error: Error solving the challenge. 'dict' object has no attribute 'decode'
2023-07-12 00:11:48 DEBUG    ReqId 5116 Response => POST /v1 body: {'status': 'error', 'message': "Error: Error solving the challenge. 'dict' object has no attribute 'decode'", 'startTimestamp': 1689117108370, 'endTimestamp': 1689117108372, 'version': '3.2.1'}
2023-07-12 00:11:48 INFO     ReqId 5116 Response in 0.002 s
2023-07-12 00:11:48 INFO     ReqId 5116 127.0.0.1 POST http://127.0.0.1:8191/v1 500 Internal Server Error
```

**Comment by ilike2burnthing** [comment]
I updated the checkbox XPATH value in flaresolverr_service.py to `//*[@id="challenge-stage"]/div/label/input` which now results in:

```text
2023-07-12 01:42:16 DEBUG    ReqId 1192 Waiting for title (attempt 4): Just a moment...
2023-07-12 01:42:26 DEBUG    ReqId 1192 Timeout waiting for selector
2023-07-12 01:42:26 DEBUG    ReqId 1192 Try to find the Cloudflare verify checkbox
2023-07-12 01:42:27 DEBUG    ReqId 1192 Cloudflare verify checkbox found and clicked
2023-07-12 01:42:27 DEBUG    ReqId 1192 Try to find the Cloudflare 'Verify you are human' button
2023-07-12 01:42:27 DEBUG    ReqId 1192 The Cloudflare 'Verify you are human' button not found on the page
```

Repeats until it times out. Commenting out the button block doesn't solve it. I'm assuming the issue is with ActionChains, but I have no idea what.

@furdarius https://github.com/FlareSolverr/FlareSolverr/commit/fc6d2d9095fa0502531fbf9452750f090b573a7d was from your work, any ideas?

**Comment by jacobprice808** [comment]
@ilike2burnthing So, I'm just using FlareSolverr to hit endpoints protected by Cloudflare. Let me see if I can replicate your issue and get to the bottom of it.

**Comment by ilike2burnthing** [comment]
The specific URLs being used by Jackett for the three indexers mentioned above are:

- https://torrentqq264.com/torrent/newest.html
- https://torrentcore.xyz/index
- https://idope.se/browse.html

**Comment by jacobprice808** [comment]
@ilike2burnthing Regarding your last comment, I think that it must be a Jackett issue because when I hit one of those IP address testing endpoints it returns my proxy IP. Here is the endpoint: https://api-bdc.net/data/client-ip

**Comment by jacobprice808** [comment]
@ilike2burnthing I'm getting the same issue when I try to hit those urls:
{"status": "error", "message": "Error: Error solving the challenge. Timeout after 60.0 seconds.", "startTimestamp": 1689127851040, "endTimestamp": 1689127911964, "version": "3.2.1"}

It is hitting the invisible Cloudflare authentication fine, though, so I think it is just struggling to find that Cloudflare element. It is still working with other versions of FlareSolverr? Is it possible that Cloudflare has changed something on its end? 

Also, if I understand correctly, FlareSolverr is having an issue getting the user agent? I'm running it fully booted on Docker and it's working fine for me for these types of URLs: https://api.prizepicks.com/leagues

I just test everything live on Docker. I made a bash script that updates the script. You might try just booting it up all the way and see if that fixes the last two problems. Here is my update script:

#!/bin/bash

cd ~

old_container_id=$(docker ps -a | grep flaresolverr | awk '{print $1}')
if [ ! -z "$old_container_id" ]; then
docker stop $old_container_id
docker rm $old_container_id
fi

if [ -d "FlareSolverr" ]; then
rm -rf FlareSolverr
fi
git clone https://github.com/jacobprice808/FlareSolverr.git
cd FlareSolverr
docker build -t myflaresolverr .

docker run -d -p 8191:8191 myflaresolverr

**Comment by ilike2burnthing** [comment]
Yea, something definitely changed with Cloudflare recently; a browser extension I use started conflicting with a few sites about a week or so ago, specifically the Cloudflare human verification challenge page. However, up until a few hours ago, FlareSolverr didn't have any issue.

I'm not seeing any Cloudflare challenge/check page when accessing https://api.prizepicks.com/leagues. Maybe it's IP dependent? I tried a few VPN servers from different countries though.

**Comment by jacobprice808** [comment]
@ilike2burnthing Does any other version of FlareSolverr work? If they're dynamically loading the element and randomizing element names and whatnot it might get tricky. Although, I just inspected the page and it doesn't seem to be doing anything crazy with that checkbox. I'm assuming you're also looking at this function in flaresolverr_service.py:
```

def click_verify(driver: WebDriver):
    try:
        logging.debug("Try to find the Cloudflare verify checkbox")
        iframe = driver.find_element(By.XPATH, "//iframe[@title='Widget containing a Cloudflare security challenge']")
        driver.switch_to.frame(iframe)
        checkbox = driver.find_element(
            by=By.XPATH,
            value='//*[@id="cf-stage"]//label[@class="ctp-checkbox-label"]/input',
        )
        if checkbox:
            actions = ActionChains(driver)
            actions.move_to_element_with_offset(checkbox, 5, 7)
            actions.click(checkbox)
            actions.perform()
            logging.debug("Cloudflare verify checkbox found and clicked")
    except Exception:
        logging.debug("Cloudflare verify checkbox not found on the page")
    finally:
        driver.switch_to.default_content()

    try:
        logging.debug("Try to find the Cloudflare 'Verify you are human' button")
        button = driver.find_element(
            by=By.XPATH,
            value="//input[@type='button' and @value='Verify you are human']",
        )
        if button:
            actions = ActionChains(driver)
            actions.move_to_element_with_offset(button, 5, 7)
            actions.click(button)
            actions.perform()
            logging.debug("The Cloudflare 'Verify you are human' button found and clicked")
    except Exception as e:
        logging.debug("The Cloudflare 'Verify you are human' button not found on the page")
        # print(e)

```

I just reread your error. I am a little bit confused. It looks like FlareSolverr is successfully bypassing CF:
```
2023-07-12 00:05:06 DEBUG    ReqId 5396 Navigating to... https://idope.se/browse.html
2023-07-12 00:05:07 INFO     ReqId 5396 Challenge detected. Title found: Just a moment...
2023-07-12 00:05:07 DEBUG    ReqId 5396 Waiting for title (attempt 4): Just a moment...
2023-07-12 00:05:18 DEBUG    ReqId 5396 Timeout waiting for selector
2023-07-12 00:05:18 DEBUG    ReqId 5396 Try to find the Cloudflare verify checkbox
2023-07-12 00:05:18 DEBUG    ReqId 5396 Cloudflare verify checkbox found and clicked
2023-07-12 00:05:18 DEBUG    ReqId 5396 Try to find the Cloudflare 'Verify you are human' button
2023-07-12 00:05:18 DEBUG    ReqId 5396 The Cloudflare 'Verify you are human' button not found on the page
```
Where is it getting stopped up?

Also, I might be overcomplicating things with my application of how I'm scraping this endpoint. I just kept getting served a Cloudflare page and decided to look into this solution.

**Comment by ilike2burnthing** [comment]
Just tried all the versions back to v3.0.4, when the `Verify you are human` code you quoted above was added - they all timeout.

Yep, it looks like it's successful, but then it makes another attempt, and another, and another, until it times out. Even with debug and HTML logging enabled, there's no sign of why it's looping like that, but I assume it's just getting another challenge page each time.

**Comment by ilike2burnthing** [comment]
https://github.com/FlareSolverr/FlareSolverr/issues/811
https://github.com/FlareSolverr/FlareSolverr/pull/812

**Comment by jacobprice808** [comment]
It might be worth it to see what's going on live in a non-headless view. Think we can recreate the issue in a Selenium + Undetected Chromedriver window view instance? I suspect that CloudFlare now has the ability to distinguish how FlareSolverr is using ActionChains and real human behavior. ActionChains are probably just going straight to the element without swerving around like a human would. Also we're going to the same pixel every time so maybe they're picking up on that? If they really are just able to distinguish these behaviors and it's not just some bug, then we should maybe try to introduce some randomness in the mouse movements. We could look into implementing something like this: https://github.com/vincentbavitz/bezmouse

**Comment by ilike2burnthing** [comment]
> Also we're going to the same pixel every time so maybe they're picking up on that?

Yea, I had the same idea and tried a few different variations without luck. Looks like we have a workaround for now.

**Comment by jacobprice808** [comment]
Yeah, I've been doing some research and I don't think it's possible to simulate low-level mouse movements in Selenium. To Cloudflare, wouldn't it just appear as if the mouse is teleporting to the element?

**Comment by benkrej** [comment]
I don't think cloudflare tracks any mouse movement or such. If you open a new tab with js the captchas are getting solved without any problems. There probably is some switch or property which triggers cf. [Example](https://github.com/ultrafunkamsterdam/undetected-chromedriver/issues/1388#issuecomment-1632000503)

**Comment by ilike2burnthing** [comment]
Updated with workaround from https://github.com/FlareSolverr/FlareSolverr/pull/816.

That shouldn't have changed anything else, but I can no longer recreate the 'error on second launch' problem from before.

This should be fine, but I want to wait for @ngosang to take a look when they're free, as we're already running a hotfix build on top of 2 janky workarounds, I'd rather not add all this on top of it as well.

**Comment by ngosang** [comment]
This PR with some fixes was merged by me in 313fb2c14b9e9c28e01f322adbab3e0ebfc752b0 Thank you all!

Included in https://github.com/FlareSolverr/FlareSolverr/releases/tag/v3.3.0

**Pull request closed without merging**